picard/version.py

```diff
@@ -0,0 +1,123 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+# Copyright (C) 2019 Philipp Wolfer
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+import re
+
+
+class VersionError(Exception):
+    pass
+
+
+class Version(tuple):
+    _version_re = re.compile(r"(\d+)[._](\d+)(?:[._](\d+)[._]?(?:(dev|a|alpha|b|beta|rc|final)[._]?(\d+))?)?$")
+
+    _identifiers = {
+        'dev': 0,
+        'alpha': 1,
+        'a': 1,
+        'beta': 2,
+        'b': 2,
+        'rc': 3,
+        'final': 4
+    }
+
+    def __new__(cls, major, minor, patch, identifier='final', revision=0):
```

Comment:
This could be init if named tuple is used?

Reply:
No, actually not. init does not work with tuples at all, since they are immutable after creation. So even with nameduple we would need to use `__new__` (but only for the identifier validation). The code would change to something like:

```
class Version(namedtuple('V', 'major minor patch identifier revision')):

    def __new__(cls, *args, **kwargs):
        instance = super(Version, cls).__new__(cls, *args, **kwargs)
        if instance.identifier not in cls.valid_identifiers():
            raise VersionError("Should be either 'final', 'dev', 'alpha', 'beta' or 'rc'")
        return instance

```

I currently don't see much value in using namedtuple here. It introduces another class in between without bringing much benefit. Having the keywords available could make the code more readable at some places (e.g. `version.minor` instead of `version[1]`). I'll experiment with it and sees if is worth the trouble.